### PR TITLE
feat(debaml): Slice 7.2a-2 — test-only canonical coercion-state collector (TEST-ONLY)

### PR DESCRIPTION
**Pull request by invakid404**
Slice 7.2a-2 of the de-BAML constraint work, per `notes/slice72a-scope.md` §2 "Constraint-aware coercion state". **TEST-ONLY.** Base `707b2419`.

## What this is

`internal/debaml/coerce.go` returns `json.RawMessage`. That is the right production contract — the serving path emits bytes — but it is deliberately insufficient for constraints. Decoding those bytes again loses the named class/enum identity, the schema-order boundary, the canonical alias decision, the skipped child/union paths, and the exact node a predicate is attached to. `{"suit":"Hearts"}` cannot tell you that `suit` is the enum `Suit`, that the model wrote the alias `hearts_alias`, or that the object is the class `Hand` rather than a bare map.

This PR rebuilds exactly that as a `constraintCoercionState` tree, **during** a canonical coercion, and runs the attached `@check`/`@assert` predicates over it through the one production evaluator seam (`ConstraintValue` + `EvaluateConstraint`).

Three files, all `_test.go`:

| file | what |
|---|---|
| `internal/debaml/constraint_state_collect_test.go` | the model + the traversal collector |
| `internal/debaml/constraint_state_test.go` | the discriminating tests |
| `internal/debaml/constraint_state_seam_test.go` | the `go/ast` test-only reachability guard |

## Path taken: pure-test — production is byte-identical to `707b2419`

`jj diff --from 707b2419 --to @ --summary` is three `A …_test.go` lines and nothing else. No non-test `.go` file changed, so no embed regen applies.

Unchanged, by construction: `coerce`'s return, `checkSupported` / `checkSupportedFields` / `checkSupportedType`, `Parse`, `ParseStaticBundle`, `SupportsNativeFinalBundle`, admission, `Checked<T>` (does not exist yet), public `bamlutils`/root symbols, mapper/codegen, the wire, and every pin/tar artifact. Constraint-bearing bundles still decline under `BAML_REST_USE_DEBAML`; `constraintCoercionRun.ProductionSupport` records `checkSupported`'s verdict on every run so a fixture cannot quietly become admitted.

## How it stays honest about "the same coercion"

The collector never re-implements a canonicalization decision it could delegate. At each node it

1. calls **production `coerce`** for that `(type, value)` and keeps the bytes,
2. builds the node's `ConstraintValue` from the **schema traversal**, using production's own `matchString` / `enumMatchCandidates` / `matchesStringToString` / `coerceListChild` / `coerceMapValueChild` / `tryCastUnion` / `selectUnionArms`, and
3. requires the two to agree (`constraintStateJSONEquivalent`).

Step 3 is load-bearing: a traversal that drifts from production — wrong field order, a dropped element kept, the wrong union arm, an alias emitted instead of a canonical name — produces a different document and the collector **fails** instead of reporting a state nothing produced.

Shapes it cannot mirror exactly are **refused by name** (`errConstraintStateUnmodelled`), not approximated: class from an array or a scalar, single-field implied-key absorption, a field filled from `TypeIR::default_value`, a field filled from the map `{}` default, recursive aliases, tuples, arrows and media.

`CanonicalJSON` is an **oracle artifact for 7.2a-3** and is never an input to the state. Identity and order come only from the traversal; only a **scalar leaf** is read back from its own one-token coercion result, through a strict decoder that rejects trailing data — and the BAML value domain is chosen by the *schema* (`float` → `Float`, `int` → `Int`), so the same token `2` lands in different domains under different fields.

## The three recorded asymmetries — represented, not papered over

**1. Bare-string return skips constraints.** The root records `SkipBareStringReturn` *before* normal node evaluation: the check collection comes back empty and a FALSE `@assert` does not reject. Both levels are tested. The skip is **positive evidence**, not an absence — for each skipped predicate the state records the *counterfactual* outcome the evaluator produced for the same canonical value, so the test proves the predicate was reached, decided `false`, and still did not reject. The control is the same constrained string type, same input, same two false predicates, moved one level down into a class field: there both events fire and `AssertFailed` is set.

**2. Duplicate labels.** Two ordered events with the same label and different results. Never folded into a map — 7.2b owns the wire/last-write-wins question. `TestConstraintStateEventsAreOrderedNotFolded` pins `Events`/`Skipped`/`Children` as slices structurally, so a refactor cannot quietly collapse them.

**3. Aliases.** The enum evaluates the **canonical** variant (`this == "Hearts"` → true, `this == "hearts_alias"` → false) and the class evaluates the **canonical** field, while `Original` retains the alias metadata for the witness. The alias spelling `this.qty` is recorded as `unsupported`, not `false` — the mapping has no such key and the operator gate refuses the resulting mixed-kind comparison. Both cases carry a control proving the metadata is a real routing observation rather than a field that is always populated. **No admission follows**: production still declines every alias-bearing constrained shape (#583).

## Dispositions

`unconstrained` · `evaluated` · `skip_bare_string_return` · `unsupported_evaluator_expression` · `skipped_child_or_union_path` · `policy_declined_constrained_shape`.

A node with no predicates is `unconstrained`, not "evaluated with nothing" — that conflation is what makes an empty-event assertion vacuous. `unsupported` is a distinct outcome from `false`, keeps its `ErrConstraintUnsupported` chain, and an undecidable `@assert` is **not** an assertion failure.

## Proof integrity

- **Discriminating only.** Assertions state the exact canonical identity, entry order, disposition, and full ordered event list, mostly through one equality against a rendered form (`class:Hand{suit=enum:Suit=Hearts,bid=int:2}`). No "no error", non-nil, truthiness or `len>0`.
- **No vacuous set arms.** Every quantified assertion first pins an exact set size (`requireConstraintStateCount`, which FATALs), or tests the empty case as its explicit subject **with a non-empty control beside it**.
- **No false-green skips.** Every skip/decline arm asserts positive evidence: the counterfactual outcome, the marker node's path, the production rule that dropped it named in the reason (`ArrayItemParseError`, `MapValueParseError`, absent optional), and a control where the same constrained type does evaluate.
- **Structural, not textual, reachability.** `constraint_state_seam_test.go` re-derives the collector's declared identifiers from the AST of its own sources and walks every non-test `.go` file in the repo (367 files, 100 identifiers) for them. A rename follows the set; a deletion makes it empty and the guard *fails as vacuous*; a move into a non-test file is itself an identifier occurrence. It also refuses any guarded name that would make the scan imprecise, refuses a production/collector name collision, and carries a synthetic-caller bite-check plus a negative control over the production evaluator seam.
- **Strict decoders, no silent drops.** Leaf readback uses `DisallowUnknownFields` + `UseNumber` + an explicit EOF check; the int leaf refuses a non-exact-i64 token instead of truncating. The class entry builder refuses a duplicate canonical field name; the map entry builder refuses a duplicate input key. A non-sentinel evaluator error is a hard failure, never an event with a swallowed cause. Tallies are labelled by what they count.
- **Every test proven to bite.** 30 mutations applied to the collector, each run against the test that should catch it; all 30 failed the test. That includes three mutations aimed specifically at the **wiring** of the divergence check — reversed class entries, a kept dropped element, an alias-spelled enum — run against a test that asserts *nothing* about order, keys or values, so only the collector's own refusal could have caught them.

## The ONE narrowed invariant — a documented, temporary, ASSERTED exception

"Constraint-bearing bundles still decline" holds for every shape in this PR **except one**, and it is narrowed nowhere else.

**The gap.** `checkSupported` → `checkSupportedFields` (`internal/debaml/parse.go`) iterates `b.Enums` and `b.Classes` and **never walks `b.Target`**. A constraint declared on the RETURN TYPE itself — `function F() -> string @assert(...)`, or a constrained element/value of a target `list`/`map` — is therefore never examined. Measured at `master` = `707b2419`:

```
target = string @assert(this == "expected"),  raw = "actual"

bundle.ValidateOutput()      = <nil>
checkSupported()             = <nil>
SupportsNativeFinalBundle()  = <nil>
ParseStaticBundle()          = json="actual"  err=<nil>     // @assert is FALSE
```

So this is an **over-claim**, not silence: native serves a result stock BAML would have rejected. `FromDynamicOutputSchema` has no constraint channel (#572 ceiling), but `FromStaticDescriptor` **does** lower constraints, so the static-final route is where it is live.

**Status.** Pre-existing at the base revision. This PR is TEST-ONLY (production byte-identical) and neither introduced it nor may fix it. The repository owner granted an **explicit, documented, temporary authority exception** for exactly this target-level shape and scheduled the fix as its own separate **"decline-more" production slice**. Recorded on **#583** ([comment](https://github.com/invakid404/baml-rest/issues/583#issuecomment-5214003796)) with the measured envelope, the reachability analysis and what the fix slice must do.

**Held by an asserted tripwire, not a silent log.** An earlier revision only `t.Log`ged the verdict so that closing the gap would not turn the suite red — but a silent log is indistinguishable from a test that checks nothing, and nothing would notice if the over-claim widened. `TestKnownGap_TargetLevelConstraintAdmission` instead:

- **asserts the wrong behaviour as a fact** — all three gates admit the bundle, and `ParseStaticBundle` serves the exact JSON;
- **demonstrates the over-claim** by driving the served value's own `@assert` through the production evaluator seam and requiring it to be **false**;
- **re-proves the exception stays confined** — the same constrained string one level down in a class field is refused by all three gates, in every world;
- **FAILS the moment the gate starts declining**, with a message naming the follow-ups to delete. It is a canary for the fix, not a permanent lock.

Verified by simulating the future fix (adding `checkSupportedType(b, b.Target)` to `checkSupported` in a throwaway edit): the tripwire fails with

```
GAP CLOSED — this is the expected fix, not a regression. Delete
TestKnownGap_TargetLevelConstraintAdmission, move the bare-string fixture from
collectConstraintStateTargetLevelFixture to collectConstraintStateFixture, and drop the
narrowed-invariant note from this file's header and the PR body. Detail: checkSupported
now DECLINES a target-level constraint: bamlutils: de-BAML parser unsupported: type constraints
```

That edit was reverted; production remains byte-identical to `707b2419`.

**Everything else is unchanged.** `TestConstraintStateConstrainedBundlesAreStillRefused` hard-asserts the refusal for 8 constrained shapes (scalar field, optional field, enum field, enum declaration, class declaration, list element, map value, map key) through `checkSupported` **+ `SupportsNativeFinalBundle` + `ParseStaticBundle`, and no assertion anywhere in this PR claims that any other bundle is admitted.

## Gates

| gate | result |
|---|---|
| `gofmt -l` on the new files | clean |
| `CGO_ENABLED=0 go build ./...` | OK |
| `go vet ./...` | OK |
| `go vet -tags=integration ./...` | OK |
| `go test ./internal/debaml/...` | ok (11.8s) |
| `go test .` (root guards) | ok |
| `TestNativeDeclines_Constraints` + the rest of `boundary_decline_test.go` | PASS |
| #649 seam tests (`TestBamlprofileConstraintAPIHasNoProductionCaller`, `TestDebamlIsTheProductionEvaluatorSeam`, `TestBamlprofileConstraintAPIListIsComplete`, `TestSeamGuardRejectsADotImportBypass`) | PASS |
| #661 guard-ledger tests + `go test -tags=integration ./internal/debaml/guardledger/` | ok (46.0s) |
| `go test -tags=integration ./internal/debaml/constraintoracle/` | ok (45.6s) |
| constraint differential case count | **719**, unchanged (0 failures) |
| production diff vs `707b2419` | **EMPTY** (three new `_test.go` files only) |
| embed regen | n/a — no non-test `.go` entered the embed set; `TestProductionBamlprofileIsEmbedded` / `TestProfileOracleIsNotEmbedded` PASS |

## Not in this PR

The serving-shaped stock oracle, the in-memory `.baml` project, stock CFFI readback and the full class/enum/list/map/asymmetry corpus are **7.2a-3**. `Checked<T>`, the carrier, mapper/codegen and the wire shape are **7.2b**. The admission decision is **7.2c**.

Two ordering facts this slice records as a stated convention rather than a proven one, both for 7.2a-3's oracle to confirm against stock: the relative order of declaration-site vs type-node constraints (recorded declaration-first, with `Origin` on every event so the split stays observable), and that the collector records the **complete** ordered event list rather than modelling any early abort on a failed `@assert`.

<a href="https://cubic.dev/pr/invakid404/baml-rest/pull/662?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Tests**
  * Expanded validation of constraint coercion across nested values, aliases, unions, collections, and optional fields.
  * Added checks for canonical value reconstruction, schema paths, ordering, skipped constraints, defaults, and unsupported inputs.
  * Added safeguards ensuring validation logic remains isolated from production behavior.
  * Improved diagnostics for mismatches, exact numeric values, invalid scalar inputs, duplicate entries, and evaluator errors.
  * Added coverage for known constraint-admission gaps and refusal scenarios.

### CODE REVIEW

File: internal/debaml/constraint_state_test.go [hunk 1]
```diff
@@ -0,0 +1,1400 @@
...
+	})
+	return t
+}
+
+func constraintStateAssert(t schema.Type, label, expr string) schema.Type {
+	t.Meta.Constraints = append(t.Meta.Constraints, schema.Constraint{
+		Level: schema.ConstraintAssert, Expression: expr, Label: &label,
+	})
+	return t
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go:162]: Fixed in `8e9c3bfae52a`. Added `constraintStateAppendConstraint`, which copies `Meta.Constraints` before appending; both `constraintStateCheck` and `constraintStateAssert` route through it.

Pinned by `TestConstraintStateFixtureAppendersDoNotAliasSiblings`. One detail worth recording: a base built by the helpers themselves always has `cap == len`, so a bare `append` allocates a fresh array anyway and a test using such a base would be vacuous — the same trap as `constraintStatePath.descend`. The fixture therefore constructs a base with **spare capacity** (asserted, so the case cannot silently disappear) and reads the first descent *after* the second is built. Verified by mutation: reverting the appender to the aliasing form fails the test.

File: internal/debaml/constraint_state_test.go [hunk 2]
```diff
@@ -0,0 +1,1400 @@
...
+		{"dropped-entry", `{"b":1}`, `{"b":1,"a":2}`, false},
+		{"reordered-array", `[2,1]`, `[1,2]`, false},
+		{"dropped-element", `[1]`, `[1,2]`, false},
+		{"different-scalar", `{"b":1}`, `{"b":2}`, false},
+		{"different-kind", `"1"`, `1`, false},
+		{"nested-reorder", `{"o":{"a":1,"b":2}}`, `{"o":{"b":2,"a":1}}`, false},
+		// The two normalizations, and ONLY these two: HTML escaping (production
+		// disables it, encoding/json does not) and float exponent spelling.
+		{"html-escaping", `"<a>"`, `"<a>"`, true},
+		{"float-exponent-spelling", `1e-7`, `1e-07`, true},
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Fixed in `8e9c3bfae52a` — the row now reads `{state: \`"\\u003ca\\u003e"\`, product: \`"<a>"\`}`, which is what the two encoders actually emit.

Went further than the row itself, since a vacuous *positive* case cannot be caught by its own outcome: both comparator tables now run `constraintStateRequireDistinctOperands`, which requires every row except the ONE explicitly marked `sameBytesIsThePoint` to feed two DIFFERENT documents, and requires exactly one such marked row. Verified by mutation — restoring the identical operands now fails with `html-escaping: operands are byte-identical ("<a>"), so the row exercises no comparison`.

File: internal/debaml/constraint_state_test.go [hunk 3]
```diff
@@ -0,0 +1,1400 @@
...
+		{"renamed-key", `{"amount":3}`, `{"qty":3}`, false},
+		{"dropped-entry", `{"b":1}`, `{"b":1,"a":2}`, false},
+		{"reordered-array", `[2,1]`, `[1,2]`, false},
+		{"dropped-element", `[1]`, `[1,2]`, false},
+		{"different-scalar", `{"b":1}`, `{"b":2}`, false},
+		{"different-kind", `"1"`, `1`, false},
+		{"nested-reorder", `{"o":{"a":1,"b":2}}`, `{"o":{"b":2,"a":1}}`, false},
+		// The two normalizations, and ONLY these two: HTML escaping (production
+		// disables it, encoding/json does not) and float exponent spelling.
+		{"html-escaping", `"<a>"`, `"<a>"`, true},
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Fixed in `8e9c3bfae52a` — same finding as the CodeRabbit thread above; thanks for the independent catch.

The row now feeds `"\\u003ca\\u003e"` (encoding/json default) against `"<a>"` (marshalJSON with `SetEscapeHTML(false)`), mirroring the float-exponent sibling. Both comparator tables additionally gained a structural non-vacuity gate requiring distinct operands on every row except the single marked byte-equality subject, so this class cannot recur silently.

File: internal/debaml/constraint_state_collect_test.go [hunk 4]
```diff
@@ -0,0 +1,1279 @@
...
+		if _, dup := seen[f.Name.Name]; dup {
+			// A schema with two identically-named fields would silently lose one
+			// through any map-shaped builder. Refuse instead.
+			return ConstraintValue{}, nil, fmt.Errorf("debaml: constraint-state collector: %s: class %q declares field %q twice", path, t.Name, f.Name.Name)
+		}
+		seen[f.Name.Name] = struct{}{}
+		entries = append(entries, ConstraintEntry{Key: f.Name.Name, Value: cs.Canonical})
+		if observed[j] != f.Name.Name || f.Name.RenderedName() != f.Name.Name {
+			st.Original.FieldAliases = append(st.Original.FieldAliases, constraintStateAliasOrigin{
+				Canonical: f.Name.Name,
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_collect_test.go]: Fixed in `8e9c3bfae52a`. The `seen` duplicate check now runs at the TOP of the per-field loop, before the default/implied/recursive descent, so a duplicate canonical field name is refused before any child is coerced or evaluated.

Verified by mutation: moving the guard back after the recursion fails `TestConstraintStateDuplicateClassFieldNamesAreRefused`.

File: internal/debaml/constraint_state_test.go [hunk 5]
```diff
@@ -0,0 +1,1846 @@
...
+	}
+	if uv.Original.Union == nil {
+		t.Fatal("defaulted union $.v recorded no union origin")
+	}
+	if got, want := *uv.Original.Union, (constraintStateUnionOrigin{Index: 1, NullArm: false, Variants: 2}); got != want {
+		t.Errorf("defaulted union $.v origin = %+v, want %+v", got, want)
+	}
+	if got, want := *uv.Original.DefaultFill, (constraintStateDefaultOrigin{Rule: "DefaultFromNoValue", ObservedKind: ""}); got != want {
+		t.Errorf("defaulted union $.v default origin = %+v, want %+v", got, want)
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go:1445]: Fixed in `8e9c3bfae52a` — `uv.Original.DefaultFill` is now nil-checked with a `t.Fatal` before the dereference, matching the sibling checks.

Swept the same class across the file while I was there: two `*v2.Original.Union` dereferences in the union control blocks had the same exposure and are now guarded too. The two `*x.Original.DefaultFill` uses inside `if x.Original.DefaultFill != nil` branches were already safe.

File: internal/debaml/constraint_state_test.go [hunk 6]
```diff
@@ -0,0 +1,1966 @@
...
+		diff, ok := constraintStateJSONEquivalent([]byte(c.state), []byte(c.product))
+		if ok != c.wantEquivalent {
+			t.Errorf("%s: equivalent=%v (want %v) for %s vs %s, diff=%q", c.name, ok, c.wantEquivalent, c.state, c.product, diff)
+		}
+	}
+	// Non-vacuity for the pairs above: they really are the same float64, so a
+	// float64 comparison WOULD have fused them.
+	if float64(9007199254740993) != float64(9007199254740992) {
+		t.Error("fixture is stale: the two tokens are no longer the same float64, so the regression case proves nothing")
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Fixed in `8e9c3bfae52a` — correct, both conversions were untyped constants and the compiler decided the comparison at compile time, so the guard was dead code.

The values now go through `int64` **variables** (`lo`, `hi`), with an added `lo == hi` fatal so a fixture collapsed to a single value is caught too. Verified by mutation: changing `hi` to a value that no longer shares a float64 with `lo` now fails the test, which it could not do before.

File: internal/debaml/constraint_state_test.go [hunk 7]
```diff
@@ -0,0 +1,1966 @@
...
+	boardValue := constraintStateClass("Board", constraintStateField("scores", "", schema.Type{
+		Kind:  schema.TypeMap,
+		Key:   ptrConstraintStateType(constraintStateStringType()),
+		Value: ptrConstraintStateType(constraintStateCheck(constraintStateIntType(), "c", "this > 0")),
+	}))
+	boardKey := constraintStateClass("Board", constraintStateField("scores", "", schema.Type{
+		Kind:  ptrConstraintStateMapKind(),
+		Key:   ptrConstraintStateType(constraintStateCheck(constraintStateStringType(), "c", `this == "a"`)),
+		Value: ptrConstraintStateType(constraintStateIntType()),
+	}))
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go:1848]: Done in `8e9c3bfae52a` — agreed, the `ptr` prefix was misleading for a value-returning helper and the indirection hid the fixture shape. `ptrConstraintStateMapKind` is deleted and the one call site now writes `Kind: schema.TypeMap` like its five siblings.

File: internal/debaml/constraint_state_test.go [hunk 8]
```diff
@@ -0,0 +1,2123 @@
...
+				if p.skipRoute {
+					verdict = "~would-be-false"
+				}
+				wantRecorded = append(wantRecorded, fmt.Sprintf("%s/%s/%s/%s%s",
+					constraintOriginTypeMeta, c.Level, label, c.Expression, verdict))
+			}
+
+			// 6. THE OVER-CLAIM ITSELF: the bundle's OWN predicate, read from the
+			// schema node above, is FALSE over the value native served.
+			requireConstraintStateCount(t, len(p.evalPaths), len(p.evalPaths), "eval paths")
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Fixed in `8e9c3bfae52a` — correct, and it was the more serious half of the finding: with `evalPaths` emptied the loop ran zero times and steps 6-7 proved nothing.

The evaluated set is no longer driven by the fixture field. It is DERIVED from the served state — every collected node whose schema type declares exactly the constraints pinned in link 1 — and the fixture field is now an *expectation* (`wantEvalPaths`) compared against that derived set. Emptying it therefore makes the coverage assertion FAIL instead of skipping the loop.

Verified by mutation, all biting: nil both probes' paths; nil only the list probe's; drop `$[1]` from the list coverage; and the worst case — delete both probe constraints AND nil every expectation AND disable the \@assert-level guard — which still fails with `constrained nodes in the served state: got 1, want 0`.

File: internal/debaml/constraint_state_test.go [hunk 9]
```diff
@@ -0,0 +1,1400 @@
...
+		Value: ptrConstraintStateType(constraintStateIntType()),
+	}
+	b := constraintStateBundle(t, constraintStateCheck(mapT, "c", "this.a == 1"), nil, nil)
+	_, err := collectConstraintCoercionState(b, `{"a":1,"a":2}`)
+	if err == nil {
+		t.Fatal("collector produced a state for a map with a duplicate input key")
+	}
+	if !strings.Contains(err.Error(), "duplicate key") {
+		t.Errorf("error = %v; want coerce_map's duplicate-key decline", err)
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Superseded — addressed by `8e9c3bfae52a`. The duplicate-key path this thread was about has been restructured since: the duplicate-map-key probe now drives the map DIRECTLY as the target (nested under a class field, `coerce_class` reclassifies `coerce_map`'s decline as an indeterminate field and the reason never surfaces), and the class duplicate-field refusal moved ahead of the recursive descent. Both are covered by `TestConstraintStateDuplicateMapKeysNeverReachTheBuilder` and `TestConstraintStateDuplicateClassFieldNamesAreRefused`, each mutation-verified.

File: internal/debaml/constraint_state_test.go [hunk 10]
```diff
@@ -0,0 +1,1966 @@
...
+}
+
+// requireConstraintStateStillDeclines asserts the boundary invariant in the ONE
+// direction that is safe to assert: a constraint-bearing bundle must still be
+// refused.
+//
+// IT IS DELIBERATELY ONE-DIRECTIONAL. An assertion that some bundle is ACCEPTED
+// would turn a gate gap into a regression lock and make closing that gap fail
+// the suite — see [TestConstraintStateConstrainedBundlesAreStillRefused], which
+// records the one shape the gate does not reach today WITHOUT pinning it open.
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_test.go]: Superseded — addressed by `8e9c3bfae52a`. All three cross-references now name `TestKnownGap_TargetLevelConstraintAdmission`:

- `requireConstraintStateStillDeclines` — rewritten; no longer claims the gap is recorded "WITHOUT pinning it open".
- the bare-string fixture comment — the dangling `TestConstraintStateSupportGateDoesNotWalkTheTargetType` reference (which never existed after the rename) is gone.
- `TestConstraintStateEveryCollectorFixtureFamilyIsRefused` — no longer refers to a "reported probe".

A grep for the superseded phrasings now returns only the deliberate historical note inside the tripwire's own doc comment.

File: internal/debaml/constraint_state_seam_test.go [hunk 11]
```diff
@@ -0,0 +1,506 @@
...
+	}
+	for _, n := range productionNames {
+		if constraintStateHasReservedPrefix(n) {
+			t.Errorf("the reserved namespace swallows the production evaluator name %q", n)
+		}
+		// And each really is declared by production, so the control is about real
+		// names rather than strings nobody uses.
+		if len(index[n]) == 0 {
+			t.Errorf("%q is not declared anywhere in internal/debaml; the negative control is stale", n)
+		}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/constraint_state_seam_test.go]: Fixed in `87d945aa8b10`. Correct and worth catching — the declaration index deliberately covers `_test.go` files too (rule 1 needs both, to tell a test-only anchor from one that leaked into production), so a presence-only check kept the negative control green for a name that had moved into test-only code and no longer represented the production evaluator seam at all.

The control now goes through a new `constraintStateDeclaredByProduction(index, name)`, which requires at least one declaration whose path does **not** end in `_test.go`, and reports the declaring files when it fails:

```
"collectConstraintStateFixture" has no PRODUCTION declaration in internal/debaml
(declared in [internal/debaml/constraint_state_test.go]); the negative control is
stale — it no longer names the production evaluator seam
```

Verified by mutation, both directions:

| | mutation | result |
|---|---|---|
| N1 | control name declared only in `_test.go` (`collectConstraintStateFixture`) | **FAILS** ✅ |
| N2 | control name declared only in `_test.go` (`requireConstraintStateCount`) | **FAILS** ✅ |
| N3 | control name declared nowhere at all | **FAILS** ✅ |
| N4 | swap in other genuinely production-declared names (`RenderConstraintExpression`, `ConstraintEntry`) | **stays green** ✅ — no false positive |
| N5 | revert to the presence-only check *and* use a test-only control name | **stays green** — i.e. exactly the blindness you reported, now demonstrated |

N5 is the counterfactual that shows the new check is load-bearing rather than decorative.

Gates on the fix: `gofmt` clean · `CGO_ENABLED=0 go build ./...` OK · `go vet ./...` and `-tags=integration` OK · `go test ./internal/debaml/...` ok · boundary + ledger + collector + known-gap 46 PASS / 0 FAIL · #649 seam + embed 6 PASS · `guardledger` and `constraintoracle` integration ok, **719 differential cases / 0 failures, unchanged** · production diff vs `707b2419` **EMPTY**.
